File: madori-tracer/renderer.py
"""構造化 JSON → 白黒間取り図 SVG を生成する。"""
from __future__ import annotations

SIZE = 800


def _n(v: float) -> float:
    return v / 100 * SIZE


def _pt(xy: list) -> str:
    return f"{_n(xy[0]):.1f},{_n(xy[1]):.1f}"


def _equipment(eq: dict) -> str:
    x, y = _n(eq["x"]), _n(eq["y"])
    w, h = _n(eq["w"]), _n(eq["h"])
    cx, cy = x + w / 2, y + h / 2
    rot = eq.get("rot", 0)
    tf = f' transform="rotate({rot},{cx:.1f},{cy:.1f})"' if rot else ""
    t = eq.get("type", "")

    if t == "toilet":
        return (
            f'<g{tf}>'
            f'<rect x="{x:.1f}" y="{y:.1f}" width="{w:.1f}" height="{h:.1f}" rx="2" fill="white" stroke="black" stroke-width="1.5"/>'
            f'<ellipse cx="{cx:.1f}" cy="{y+h*0.65:.1f}" rx="{w*0.38:.1f}" ry="{h*0.28:.1f}" fill="white" stroke="black" stroke-width="1.2"/>'
            f'<rect x="{x+w*0.1:.1f}" y="{y+h*0.05:.1f}" width="{w*0.8:.1f}" height="{h*0.22:.1f}" rx="2" fill="white" stroke="black" stroke-width="1"/>'
            f'</g>'
        )
    if t == "bath":
        return (
            f'<g{tf}>'
            f'<rect x="{x:.1f}" y="{y:.1f}" width="{w:.1f}" height="{h:.1f}" rx="3" fill="white" stroke="black" stroke-width="1.5"/>'
            f'<ellipse cx="{cx:.1f}" cy="{cy+h*0.08:.1f}" rx="{w*0.37:.1f}" ry="{h*0.33:.1f}" fill="white" stroke="black" stroke-width="1.2"/>'
            f'<rect x="{x+w*0.1:.1f}" y="{y+h*0.05:.1f}" width="{w*0.8:.1f}" height="{h*0.1:.1f}" rx="1" fill="#ccc" stroke="black" stroke-width="0.8"/>'
            f'</g>'
        )
    if t == "sink":
        r = min(w, h) * 0.32
        return (
            f'<g{tf}>'
            f'<rect x="{x:.1f}" y="{y:.1f}" width="{w:.1f}" height="{h:.1f}" rx="2" fill="white" stroke="black" stroke-width="1.5"/>'
            f'<ellipse cx="{cx:.1f}" cy="{cy:.1f}" rx="{r:.1f}" ry="{r:.1f}" fill="white" stroke="black" stroke-width="1.2"/>'
            f'<circle cx="{cx:.1f}" cy="{cy:.1f}" r="{min(w,h)*0.07:.1f}" fill="black"/>'
            f'</g>'
        )
    if t == "kitchen_sink":
        pw = w * 0.43
        ph = h * 0.76
        return (
            f'<g{tf}>'
            f'<rect x="{x:.1f}" y="{y:.1f}" width="{w:.1f}" height="{h:.1f}" fill="white" stroke="black" stroke-width="1.5"/>'
            f'<rect x="{x+w*0.05:.1f}" y="{y+h*0.12:.1f}" width="{pw:.1f}" height="{ph:.1f}" rx="2" fill="white" stroke="black" stroke-width="1"/>'
            f'<rect x="{x+w*0.52:.1f}" y="{y+h*0.12:.1f}" width="{pw:.1f}" height="{ph:.1f}" rx="2" fill="white" stroke="black" stroke-width="1"/>'
            f'</g>'
        )
    if t == "stove":
        r = min(w, h) * 0.15
        return (
            f'<g{tf}>'
            f'<rect x="{x:.1f}" y="{y:.1f}" width="{w:.1f}" height="{h:.1f}" fill="white" stroke="black" stroke-width="1.5"/>'
            f'<circle cx="{x+w*0.27:.1f}" cy="{cy:.1f}" r="{r:.1f}" fill="white" stroke="black" stroke-width="1"/>'
            f'<circle cx="{x+w*0.73:.1f}" cy="{cy:.1f}" r="{r:.1f}" fill="white" stroke="black" stroke-width="1"/>'
            f'<circle cx="{cx:.1f}" cy="{y+h*0.28:.1f}" r="{r*0.8:.1f}" fill="white" stroke="black" stroke-width="1"/>'
            f'</g>'
        )
    if t == "washing_machine":
        r = min(w, h) * 0.33
        return (
            f'<g{tf}>'
            f'<rect x="{x:.1f}" y="{y:.1f}" width="{w:.1f}" height="{h:.1f}" rx="2" fill="white" stroke="black" stroke-width="1.5"/>'
            f'<circle cx="{cx:.1f}" cy="{cy:.1f}" r="{r:.1f}" fill="white" stroke="black" stroke-width="1.2"/>'
            f'<circle cx="{cx:.1f}" cy="{cy:.1f}" r="{r*0.3:.1f}" fill="#ddd" stroke="black" stroke-width="0.8"/>'
            f'</g>'
        )
    return f'<rect x="{x:.1f}" y="{y:.1f}" width="{w:.1f}" height="{h:.1f}" fill="white" stroke="black" stroke-width="1.2"{tf}/>'


def _door(d: dict) -> str:
    x, y = _n(d["x"]), _n(d["y"])
    w = _n(d["w"])
    rot = d.get("rot", 0)
    swing = d.get("swing", "left")
    tf = f'rotate({rot},{x:.1f},{y:.1f})'
    if swing == "right":
        arc = f'<path d="M {x+w:.1f} {y:.1f} A {w:.1f} {w:.1f} 0 0 1 {x:.1f} {y+w:.1f}" fill="none" stroke="black" stroke-width="1" stroke-dasharray="4,2"/>'
    else:
        arc = f'<path d="M {x:.1f} {y:.1f} A {w:.1f} {w:.1f} 0 0 0 {x+w:.1f} {y+w:.1f}" fill="none" stroke="black" stroke-width="1" stroke-dasharray="4,2"/>'
    line = f'<line x1="{x:.1f}" y1="{y:.1f}" x2="{x+w:.1f}" y2="{y:.1f}" stroke="black" stroke-width="2"/>'
    return f'<g transform="{tf}">{line}{arc}</g>'


def _window(win: dict) -> str:
    x, y = _n(win["x"]), _n(win["y"])
    w = _n(win["w"])
    rot = win.get("rot", 0)
    tf = f' transform="rotate({rot},{x:.1f},{y:.1f})"' if rot else ""
    g = 2.5
    return (
        f'<g{tf}>'
        f'<line x1="{x:.1f}" y1="{y-g:.1f}" x2="{x+w:.1f}" y2="{y-g:.1f}" stroke="black" stroke-width="1"/>'
        f'<line x1="{x:.1f}" y1="{y:.1f}" x2="{x+w:.1f}" y2="{y:.1f}" stroke="black" stroke-width="3"/>'
        f'<line x1="{x:.1f}" y1="{y+g:.1f}" x2="{x+w:.1f}" y2="{y+g:.1f}" stroke="black" stroke-width="1"/>'
        f'</g>'
    )


def _compass(cx: float, cy: float) -> str:
    r = 18
    return (
        f'<circle cx="{cx:.1f}" cy="{cy:.1f}" r="{r}" fill="white" stroke="black" stroke-width="1.5"/>'
        f'<polygon points="{cx:.1f},{cy-r+3:.1f} {cx-5:.1f},{cy+4:.1f} {cx+5:.1f},{cy+4:.1f}" fill="black"/>'
        f'<text x="{cx:.1f}" y="{cy+r-4:.1f}" text-anchor="middle" '
        f'font-size="10" font-weight="bold" fill="black">N</text>'
    )


def render_svg(data: dict) -> str:
    parts: list[str] = []

    parts.append(f'<rect width="{SIZE}" height="{SIZE}" fill="white"/>')

    # 外形（外壁の内側を白塗り）
    outer = data.get("outer_polygon", [])
    if outer:
        pts = " ".join(_pt(p) for p in outer)
        parts.append(f'<polygon points="{pts}" fill="white" stroke="none"/>')

    # 部屋ラベル
    for room in data.get("rooms", []):
        rx, ry = _n(room["x"]), _n(room["y"])
        rw, rh = _n(room["w"]), _n(room["h"])
        cx, cy = rx + rw / 2, ry + rh / 2
        name = room.get("name", "")
        size = room.get("size", "")
        fs = 10 if len(name) > 7 else (12 if len(name) > 4 else 14)
        parts.append(
            f'<text x="{cx:.1f}" y="{cy - (6 if size else 0):.1f}" text-anchor="middle" '
            f'font-size="{fs}" font-weight="bold" fill="black">{name}</text>'
        )
        if size:
            parts.append(
                f'<text x="{cx:.1f}" y="{cy+12:.1f}" text-anchor="middle" '
                f'font-size="11" fill="#222">{size}</text>'
            )

    # 柱（黒塗り）— 壁より前に描く
    for p in data.get("pillars", []):
        parts.append(
            f'<rect x="{_n(p["x"]):.1f}" y="{_n(p["y"]):.1f}" '
            f'width="{_n(p["w"]):.1f}" height="{_n(p["h"]):.1f}" fill="black"/>'
        )

    # 窓
    for win in data.get("windows", []):
        parts.append(_window(win))

    # 外壁
    if outer:
        pts = " ".join(_pt(p) for p in outer)
        parts.append(
            f'<polygon points="{pts}" fill="none" stroke="black" stroke-width="5" stroke-linejoin="miter"/>'
        )

    # 内壁
    for wall in data.get("inner_walls", []):
        f = wall["from"]
        t = wall["to"]
        parts.append(
            f'<line x1="{_n(f[0]):.1f}" y1="{_n(f[1]):.1f}" '
            f'x2="{_n(t[0]):.1f}" y2="{_n(t[1]):.1f}" '
            f'stroke="black" stroke-width="2.5" stroke-linecap="square"/>'
        )

    # 設備
    for eq in data.get("equipment", []):
        parts.append(_equipment(eq))

    # ドア
    for door in data.get("doors", []):
        parts.append(_door(door))

    # ラベル（PS・MB等）
    for lbl in data.get("labels", []):
        parts.append(
            f'<text x="{_n(lbl["x"]):.1f}" y="{_n(lbl["y"]):.1f}" text-anchor="middle" '
            f'font-size="9" fill="black">{lbl["text"]}</text>'
        )

    # 方位磁針
    comp = data.get("compass", {"x": 92, "y": 92})
    parts.append(_compass(_n(comp["x"]), _n(comp["y"])))

    font = "'Noto Sans JP','Yu Gothic','Hiragino Kaku Gothic ProN',sans-serif"
    inner = "\n  ".join(parts)
    return f"""<?xml version="1.0" encoding="UTF-8"?>
<svg xmlns="http://www.w3.org/2000/svg" width="{SIZE}" height="{SIZE}" viewBox="0 0 {SIZE} {SIZE}">
  <defs>
    <style>text {{ font-family: {font}; }}</style>
  </defs>
  {inner}
</svg>"""

File: madori-tracer/lib/renderer.ts
import { FloorPlanJSON, Equipment, Door, Window } from "./types";

const SVG_SIZE = 800; // output SVG canvas size

function norm(v: number) {
  return (v / 100) * SVG_SIZE;
}

function pt([x, y]: [number, number]) {
  return `${norm(x)},${norm(y)}`;
}

// Equipment SVG icons (simplified line art)
function renderEquipment(eq: Equipment): string {
  const x = norm(eq.x);
  const y = norm(eq.y);
  const w = norm(eq.width);
  const h = norm(eq.height);
  const cx = x + w / 2;
  const cy = y + h / 2;
  const transform = eq.rotation ? ` transform="rotate(${eq.rotation},${cx},${cy})"` : "";

  switch (eq.type) {
    case "toilet":
      return `<g${transform}>
        <rect x="${x}" y="${y}" width="${w}" height="${h}" rx="2" fill="white" stroke="black" stroke-width="1.5"/>
        <ellipse cx="${cx}" cy="${y + h * 0.65}" rx="${w * 0.38}" ry="${h * 0.28}" fill="white" stroke="black" stroke-width="1.2"/>
        <rect x="${x + w * 0.1}" y="${y + h * 0.05}" width="${w * 0.8}" height="${h * 0.25}" rx="2" fill="white" stroke="black" stroke-width="1"/>
      </g>`;

    case "bath":
      return `<g${transform}>
        <rect x="${x}" y="${y}" width="${w}" height="${h}" rx="3" fill="white" stroke="black" stroke-width="1.5"/>
        <ellipse cx="${cx}" cy="${cy + h * 0.1}" rx="${w * 0.38}" ry="${h * 0.35}" fill="white" stroke="black" stroke-width="1.2"/>
        <rect x="${x + w * 0.1}" y="${y + h * 0.05}" width="${w * 0.8}" height="${h * 0.12}" rx="1" fill="#ddd" stroke="black" stroke-width="0.8"/>
      </g>`;

    case "sink":
      return `<g${transform}>
        <rect x="${x}" y="${y}" width="${w}" height="${h}" rx="2" fill="white" stroke="black" stroke-width="1.5"/>
        <ellipse cx="${cx}" cy="${cy}" rx="${w * 0.35}" ry="${h * 0.35}" fill="white" stroke="black" stroke-width="1.2"/>
        <circle cx="${cx}" cy="${cy}" r="${Math.min(w, h) * 0.06}" fill="black"/>
      </g>`;

    case "kitchen_sink":
      return `<g${transform}>
        <rect x="${x}" y="${y}" width="${w}" height="${h}" fill="white" stroke="black" stroke-width="1.5"/>
        <rect x="${x + w * 0.05}" y="${y + h * 0.1}" width="${w * 0.42}" height="${h * 0.75}" rx="2" fill="white" stroke="black" stroke-width="1"/>
        <rect x="${x + w * 0.53}" y="${y + h * 0.1}" width="${w * 0.42}" height="${h * 0.75}" rx="2" fill="white" stroke="black" stroke-width="1"/>
      </g>`;

    case "stove":
      return `<g${transform}>
        <rect x="${x}" y="${y}" width="${w}" height="${h}" fill="white" stroke="black" stroke-width="1.5"/>
        <circle cx="${x + w * 0.28}" cy="${cy}" r="${Math.min(w, h) * 0.18}" fill="white" stroke="black" stroke-width="1"/>
        <circle cx="${x + w * 0.72}" cy="${cy}" r="${Math.min(w, h) * 0.18}" fill="white" stroke="black" stroke-width="1"/>
        <circle cx="${cx}" cy="${y + h * 0.25}" r="${Math.min(w, h) * 0.12}" fill="white" stroke="black" stroke-width="1"/>
      </g>`;

    case "washing_machine":
      return `<g${transform}>
        <rect x="${x}" y="${y}" width="${w}" height="${h}" rx="2" fill="white" stroke="black" stroke-width="1.5"/>
        <circle cx="${cx}" cy="${cy}" r="${Math.min(w, h) * 0.35}" fill="white" stroke="black" stroke-width="1.2"/>
        <circle cx="${cx}" cy="${cy}" r="${Math.min(w, h) * 0.1}" fill="#ddd" stroke="black" stroke-width="0.8"/>
      </g>`;

    default:
      return `<rect x="${x}" y="${y}" width="${w}" height="${h}" fill="white" stroke="black" stroke-width="1.2"${transform}/>`;
  }
}

// Door arc rendering
function renderDoor(door: Door): string {
  const x = norm(door.x);
  const y = norm(door.y);
  const w = norm(door.width);
  const cx = x;
  const cy = y;
  const transform = `rotate(${door.rotation},${x},${y})`;

  // Door leaf (line) + swing arc
  const arcX = door.swing === "right" ? x + w : x;
  const sweepFlag = door.swing === "right" ? 1 : 0;
  const endX = door.swing === "right" ? x + w : x - w;

  return `<g transform="${transform}">
    <line x1="${x}" y1="${y}" x2="${arcX}" y2="${y}" stroke="black" stroke-width="2"/>
    <path d="M ${arcX} ${cy} A ${w} ${w} 0 0 ${sweepFlag} ${endX} ${y + w}"
          fill="none" stroke="black" stroke-width="1" stroke-dasharray="4,2"/>
  </g>`;
}

// Window rendering (triple line style)
function renderWindow(win: Window): string {
  const x = norm(win.x);
  const y = norm(win.y);
  const w = norm(win.width);
  const transform = win.rotation ? ` transform="rotate(${win.rotation},${x},${y})"` : "";
  const gap = 2.5;

  return `<g${transform}>
    <line x1="${x}" y1="${y - gap}" x2="${x + w}" y2="${y - gap}" stroke="black" stroke-width="1"/>
    <line x1="${x}" y1="${y}" x2="${x + w}" y2="${y}" stroke="black" stroke-width="2.5"/>
    <line x1="${x}" y1="${y + gap}" x2="${x + w}" y2="${y + gap}" stroke="black" stroke-width="1"/>
  </g>`;
}

// Compass (N mark)
function renderCompass(x: number, y: number): string {
  const r = 18;
  return `<g>
    <circle cx="${x}" cy="${y}" r="${r}" fill="white" stroke="black" stroke-width="1.5"/>
    <polygon points="${x},${y - r + 4} ${x - 5},${y + 4} ${x + 5},${y + 4}" fill="black"/>
    <text x="${x}" y="${y + r - 4}" text-anchor="middle" font-family="'Noto Sans JP', sans-serif"
          font-size="10" font-weight="bold" fill="black">N</text>
  </g>`;
}

export function renderSVG(data: FloorPlanJSON): string {
  const parts: string[] = [];

  // Background
  parts.push(`<rect width="${SVG_SIZE}" height="${SVG_SIZE}" fill="white"/>`);

  // Room fills (white) + room name labels
  for (const room of data.rooms) {
    if (room.polygon.length < 3) continue;
    const points = room.polygon.map(pt).join(" ");
    parts.push(`<polygon points="${points}" fill="white" stroke="none"/>`);

    // Room label: center of bounding box
    const xs = room.polygon.map(([x]) => norm(x));
    const ys = room.polygon.map(([, y]) => norm(y));
    const lx = (Math.min(...xs) + Math.max(...xs)) / 2;
    const ly = (Math.min(...ys) + Math.max(...ys)) / 2;
    const nameFontSize = room.name.length > 4 ? 11 : 13;

    parts.push(`<text x="${lx}" y="${ly - 6}" text-anchor="middle"
      font-family="'Noto Sans JP', sans-serif" font-size="${nameFontSize}" font-weight="bold" fill="black">${room.name}</text>`);
    if (room.size) {
      parts.push(`<text x="${lx}" y="${ly + 10}" text-anchor="middle"
        font-family="'Noto Sans JP', sans-serif" font-size="10" fill="#444">${room.size}</text>`);
    }
  }

  // Pillars (solid black)
  for (const p of data.pillars) {
    parts.push(`<rect x="${norm(p.x)}" y="${norm(p.y)}" width="${norm(p.width)}" height="${norm(p.height)}" fill="black"/>`);
  }

  // Windows
  for (const win of data.windows) {
    parts.push(renderWindow(win));
  }

  // Walls
  for (const wall of data.walls) {
    const strokeWidth = wall.type === "outer" ? 3.5 : 1.8;
    parts.push(`<line x1="${norm(wall.from[0])}" y1="${norm(wall.from[1])}"
      x2="${norm(wall.to[0])}" y2="${norm(wall.to[1])}"
      stroke="black" stroke-width="${strokeWidth}" stroke-linecap="square"/>`);
  }

  // Equipment
  for (const eq of data.equipment) {
    parts.push(renderEquipment(eq));
  }

  // Doors
  for (const door of data.doors) {
    parts.push(renderDoor(door));
  }

  // Compass
  const compassX = norm(data.compass.x);
  const compassY = norm(data.compass.y);
  parts.push(renderCompass(compassX, compassY));

  // Scale label
  if (data.scale) {
    parts.push(`<text x="10" y="${SVG_SIZE - 10}"
      font-family="'Noto Sans JP', sans-serif" font-size="10" fill="#666">縮尺 ${data.scale}</text>`);
  }

  return `<?xml version="1.0" encoding="UTF-8"?>
<svg xmlns="http://www.w3.org/2000/svg" width="${SVG_SIZE}" height="${SVG_SIZE}" viewBox="0 0 ${SVG_SIZE} ${SVG_SIZE}">
  <defs>
    <style>
      text { font-family: 'Noto Sans JP', 'Yu Gothic', 'Hiragino Kaku Gothic ProN', sans-serif; }
    </style>
  </defs>
  ${parts.join("\n  ")}
</svg>`;
}
